x/msgfees/keeper/keeper.go

```diff
@@ -130,16 +141,62 @@ func (k Keeper) IterateMsgFees(ctx sdk.Context, handle func(msgFees types.MsgFee
-// cosmos method does not take in the custom fee collector which is a feature desired from msg fees.
-func (k Keeper) DeductFees(bankKeeper cosmosauthtypes.BankKeeper, ctx sdk.Context, acc cosmosauthtypes.AccountI, fees sdk.Coins) error {
-	if !fees.IsValid() {
-		return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, "invalid fee amount: %q", fees)
+// DeductFeesDistributions deducts fees from the given account.  The fees map contains a key of bech32 addresses to distribute funds to.
+// If the key in the map is an empty string, those will go to the fee collector.  After all the accounts in fees map are paid out,
+// the remainder of remainingFees will be swept to the fee collector account.
+func (k Keeper) DeductFeesDistributions(bankKeeper bankkeeper.Keeper, ctx sdk.Context, acc cosmosauthtypes.AccountI, remainingFees sdk.Coins, fees map[string]sdk.Coins) error {
+	sentCoins := sdk.NewCoins()
+	for key, coins := range fees {
+		if !coins.IsValid() {
+			return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFee, "invalid fee amount: %q", fees)
+		}
+		if len(key) == 0 {
+			err := bankKeeper.SendCoinsFromAccountToModule(ctx, acc.GetAddress(), k.FeeCollectorName, coins)
+			if err != nil {
+				return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFunds, err.Error())
+			}
+		} else {
+			recipient, err := sdk.AccAddressFromBech32(key)
+			if err != nil {
+				return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, err.Error())
+			}
+			err = bankKeeper.SendCoins(ctx, acc.GetAddress(), recipient, coins)
+			if err != nil {
+				return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFunds, err.Error())
+			}
+		}
+		sentCoins = sentCoins.Add(coins...)
 	}
-
-	err := bankKeeper.SendCoinsFromAccountToModule(ctx, acc.GetAddress(), k.feeCollectorName, fees)
+	remainingFee, neg := remainingFees.SafeSub(sentCoins)
+	if neg {
+		return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFunds, "negative balance after sending coins to accounts and fee collector")
+	}
+	// sweep the rest of the fees to module
+	err := bankKeeper.SendCoinsFromAccountToModule(ctx, acc.GetAddress(), k.FeeCollectorName, remainingFee)
 	if err != nil {
 		return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFunds, err.Error())
 	}
+
 	return nil
 }
+
+// ConvertDenomToHash converts coin to nhash coin using conversion rate represented as dollar mill ($1.234 = 1234).
+// Currently, usd is only supported with conversion rate coming from params
+func (k Keeper) ConvertDenomToHash(ctx sdk.Context, coin sdk.Coin) (sdk.Coin, error) {
+	if coin.Denom == "usd" {
+		var nhashAmount sdk.Int
+		conversionRate := int64(k.GetUsdConversionRate(ctx))
+		if conversionRate > coin.Amount.Int64() {
+			hashAmount := conversionRate / coin.Amount.Int64()
+			nhashAmount = sdk.NewInt(100_000_000).Mul(sdk.NewInt(hashAmount))
+		} else {
+			hashAmount := coin.Amount.Int64() / conversionRate
+			nhashAmount = sdk.NewInt(1_000_000_000).Mul(sdk.NewInt(hashAmount))
+		}
```

Comment:
Okay. The conversion formula is `hash = USD.000 / rate`. Since 1 hash = 1,000,000,000 nhash, that same conversion formula is `nhash = USD.000 * 1,000,000,000 / rate.

The multiplication there needs to be done before division.
So the conversion should be done like this:
```suggestion
		nhashAmount := sdk.Int(coin.Amount.Int64() * 1_000_000_000 / conversionRate)
```

Reply:
@dwedul-figure thanks for this.  I was looking at a way to refactor it and make it simpler.  I have some tests written around it and will change it to see if they still pass.

Reply:
I didn't see any unit tests on it, but if there were already some, and they all passed, I'm concerned about them. Specifically, this part:
```go
if conversionRate > coin.Amount.Int64() {
	hashAmount := conversionRate / coin.Amount.Int64()
	nhashAmount = sdk.NewInt(100_000_000).Mul(sdk.NewInt(hashAmount))
}
```
it is wrong. The fraction can't just be inverted like that, and the conversion of hash to nhash is always 1,000,000,000.
So if there are any unit tests where conversion rate is > coin amount, and those tests pass using this calculation, those unit tests are only enforcing incorrectness.

Reply:
I also believe we should only work in nhash here and not mix in the idea of rates in hash proper (10^9 nhash).